zee/src/editor/buffer.rs

```diff
@@ -173,11 +173,16 @@ impl Buffer {
         file_path: Option<PathBuf>,
         repo: Option<RepositoryRc>,
     ) -> Self {
-        let mode = file_path
-            .as_ref()
-            .map(|path| context.0.mode_by_filename(path))
+        let mode = text
+            .line(0)
+            .as_str()
+            .and_then(|shebang| context.0.mode_by_shebang(shebang))
+            .or_else(|| {
+                file_path
+                    .as_ref()
+                    .and_then(|path| context.0.mode_by_filename(path))
+            })
```

Comment:
This may potentially read the whole file in pathological cases, e.g. a minified file that has no new lines. One goal of `zee` is to keep being fast for any kind of pathological file you can think of and do anything that is potentially blocking in the background (e.g. parsing syntax or writing the file to disk). I've also been trying to avoid doing anything linear in the length of a line in the UI thread. 

I think the right solution here long term is to build buffers, i.e. call `Buffer::new()` in a background thread, rather than in the main, UI thread.

For this PR though, I'd be happy if instead you just bound how much of the file you read, say 256 bytes at most and test the regex for that. You'll have to deal with potentially truncated utf-8...

Maybe a better solution is to read characters until you encounter either 1. a new line or 2. if you don't after X characters, you give you and don't check the shebang -- essentially we only test if shebangs apply up to a certain line length.

Reply:
A 2nd comment is that I think we should test if "mode_by_filename" applies and only then check the shebangs to avoid having to read the file if the name already matches.